estimator/ntru_primal.py

```diff
@@ -0,0 +1,503 @@
+
+    @staticmethod
+    def log_gh(d, logvol=0):
+        if d < 49:
+            return RR(gh_constant[d]) + RR(logvol)/d
+
+        return RR(1./d) * RR(logvol - PrimalDSD.ball_log_vol(d))
+
+    @staticmethod
+    def DSL_logvol_matrix(n, sigmasq):
+        total = n*(RR(log(sigmasq))+RR(log(2.))+RR(digamma(n)))/2.
+        proj_loss = np.sum([(digamma((2*n-i)/2.)-digamma(n)) for i in range(n)])/2.
+        return total+proj_loss
+
+    @staticmethod
+    def DSL_logvol_circulant(n, sigmasq):
+        lambda0 = RR((np.log(2)-np.euler_gamma+np.log(n)+np.log(sigmasq))/2.)
+        lambdai = (n-1)*(1-np.euler_gamma+np.log(n)+np.log(sigmasq))/2.
+        return lambda0+lambdai
+
+    @staticmethod
+    def DSL_logvol_circulant_fixed(n, R):
+        lambda0 = (-np.euler_gamma+np.log(R))/2.
+        lambdai = (n-1)*(1-np.euler_gamma+np.log(R)-np.log(2))/2.
+        return lambda0+lambdai
+
+    @staticmethod
+    @cached_function
+    def DSL_logvol(n, sigmasq, ntru="circulant"):
+        if ntru=="matrix":
+            return PrimalDSD.DSL_logvol_matrix(n, sigmasq)
+        if ntru=="circulant":
+            return PrimalDSD.DSL_logvol_circulant(n, sigmasq)
+        if ntru=="fixed":
+            return PrimalDSD.DSL_logvol_circulant_fixed(n, sigmasq)
+        print("non implemented ntru type")
+
+    @staticmethod
+    @cached_function
+    def proj_logloss(d, k):
+        # log loss of length when projecting out k dimension out of d
+        return (RR(digamma((d-k)/2.))-RR(digamma(d/2.)))/2.
+
+    @staticmethod
+    def DSLI_vols(dsl_logvol, FL_shape):
+        n = len(FL_shape)//2
+        vols = (2*n+1)*[None]
+
+        dsl_dim = n
+        vols[2*n] = dsl_logvol
+
+        # Going to a intersection of dimension s
+        for s in range(2*n-1, n, -1):
+            # Negate cause it's a dual vector really
+            x = - FL_shape[s]
+            x += PrimalDSD.proj_logloss(s+1, n)
+            x += zeta_prime_precomputed[dsl_dim]/zeta_precomputed[dsl_dim]  # primitivity
+            dsl_logvol += x
+            vols[s] = dsl_logvol
+            dsl_dim -= 1
+
+        assert dsl_dim == 1
+        assert s == n+1
+
+        return vols
+
+    @staticmethod
+    @cached_function
+    def prob_dsd(
+        beta: int,
+        params: NTRUParameters,
+        simulator,
+        m: int = oo,
+        tau=None,
+        d=None,
+        dsl_logvol=None,
+        red_cost_model=red_cost_model_default,
+        log_level=None,
+    ):
+
+        if d is None:
+            d = m
+
+        xi = PrimalUSVP._xi_factor(params.Xs, params.Xe)
+        if dsl_logvol is None:
+            dsl_logvol = PrimalDSD.DSL_logvol(params.n, params.Xs.stddev**2, ntru=params.ntru_type)
+
+        B_shape = [log(r_)/2 for r_ in simulator(d, params.n, params.q, beta, xi=xi, tau=tau)]
+
+        dsli_vols = PrimalDSD.DSLI_vols(dsl_logvol, B_shape)
+
+        prob_all_not = RR(1.)
+        prob_pos = np.zeros(2*params.n, dtype='double')
```

Comment:
Do we really need to use numpy here, given that we're already in Sage.

Reply:
Isn't i just a list? Also, sage matrices are mutable, no?

Reply:
Yeah. I assumed there was some vectorization, but I misread how it was being used. Changed to a list in the latest commit.